src/zjs_util.h

```diff
@@ -275,4 +275,66 @@ int zjs_validate_args(const char *expectations[], const jerry_length_t argc,
         (zjs_validate_args((const char *[]){ __VA_ARGS__, NULL }, argc, argv) \
                 <= ZJS_INVALID_ARG) ? 1 : 0
 
+#define ZJS_FIND_NODE(type, list, cmp_element, cmp_to) \
+   ({ \
+       type *found = NULL; \
+       type *cur = list; \
+       while (cur) { \
+           if (cur->cmp_element == cmp_to) { \
+               found = cur; \
+               break; \
+           } \
+           cur = cur->next; \
+       } \
+       found; \
+   })
+
+#define ZJS_APPEND_NODE(type, list, p) \
+    { \
+        type **pnext = &list; \
+        while (*pnext != NULL) { \
+            pnext = &(*pnext)->next; \
+        } \
+    *pnext = p; \
+    }
+#define ZJS_PREPEND_NODE(type, list, p) \
+    { \
+        ((type *)p)->next = list; \
+        list = p; \
+    }
+
+#define ZJS_REMOVE_NODE(type, list, p) \
+    type *cur = list; \
+    if (p == list) { \
+        list = ((type *)p)->next; \
+    } else { \
+        while (cur->next) { \
+            if (cur->next == p) { \
+                cur->next = ((type *)p)->next; \
+                break; \
+            } \
+            cur = cur->next; \
+        } \
+    }
+
+#define ZJS_FREE_LIST(type, list, callback) \
```

Comment:
Change all functions to start with ZJS_LIST prefix, you can then probably leave off _NODE.

Probably this should just call zjs_free directly instead of a callback? I assume that's your real use case?

Otherwise it's more like FOREACH except that it sets the list to NULL at the end. Maybe a FOREACH would be useful.